### ISSUE TITLE
CI: [v1.15] Cilium E2E upgrade - no unexpected packet drops ("Missed tail call")

### ISSUE THREAD

**Issue by giorio94** [cilium/cilium#35594]
## CI failure

The Cilium E2E upgrade workflow on the v1.15 branch appears to have a success rate at around 35% \[1] since having been enabled again ~3 weeks ago \[2]. From a quick look, most of the failures appear to be due to `no-unexpected-packet-drops`:

```
❌ Found unexpected packet drops:
{
  "labels": {
    "direction": "EGRESS",
    "reason": "Missed tail call"
  },
  "name": "cilium_drop_count_total",
  "value": 1
}
```

Example runs:
* https://github.com/cilium/cilium/actions/runs/11546212158/job/32134246250
* https://github.com/cilium/cilium/actions/runs/11546212158/job/32134247860
* https://github.com/cilium/cilium/actions/runs/11543410791/job/32127773250
* https://github.com/cilium/cilium/actions/runs/11528102534/job/32094779424
* https://github.com/cilium/cilium/actions/runs/11524650199/job/32085300272

\[1]: https://github.com/cilium/cilium/actions/workflows/tests-e2e-upgrade.yaml?query=event%3Aworkflow_dispatch+branch%3Av1.15
\[2]: 9bde97d05432bb62d654c51bb9b7931f751ac022

**Comment by gandro** [comment]
This also seems pretty bad on v1.16 in particular for e2e upgrade test 2: https://github.com/cilium/cilium/actions/runs/11614940433/job/32483752297

**Comment by jrajahalme** [comment]
Same problem with `ci-ipsec-upgrade`, 10 failures in a row for the same v1.15 backportPR (https://github.com/cilium/cilium/pull/35684): https://github.com/cilium/cilium/actions/runs/11642994888

**Comment by gandro** [comment]
Some more data: On v1.15, I seem to have observed two missed tail calls in bpf_xdp.c interestingly enough: 

```
REASON                                    DIRECTION   PACKETS   BYTES       LINE   FILE
Interface                                 INGRESS     1062741   331662248   1111   bpf_host.c
Interface                                 INGRESS     1186254   644052752   1190   bpf_host.c
Interface                                 INGRESS     6000      503169      1186   bpf_host.c
Interface                                 INGRESS     9         1442        670    bpf_overlay.c
Invalid source ip                         EGRESS      42        3260        808    bpf_lxc.c
Invalid source ip                         EGRESS      47        3490        778    bpf_lxc.c
Missed tail call                          INGRESS     1         150         288    bpf_xdp.c
```

https://github.com/cilium/cilium/blob/e580c18c63c428f3b0bdf6bf5a888d954f628bd9/bpf/bpf_xdp.c#L286-L288

**Comment by christarazi** [comment]
@ti-mo I recall you've been looking into missed tail calls recently, do you have any input here?

**Comment by gandro** [comment]
I went through the failures for the last week in https://github.com/cilium/cilium/actions/workflows/tests-e2e-upgrade.yaml?query=event%3Aworkflow_dispatch+branch%3Av1.15. Of the missed tail call failiures, all are of the following form:

```
{
  "labels": {
    "direction": "INGRESS",
    "reason": "Missed tail call"
  },
  "name": "cilium_drop_count_total",
  "value": 2
}
```

Direction is always INGRESS, count is 1 or 2.

In the last week, the failure only occurred in three tests:

Test 6: 5 occurrences
Test 7: 4 occurrences
Test 14: 2 occurrences

**Comment by ti-mo** [comment]
Sent a small fix for 1.14 and 1.15 that may alleviate this:
https://github.com/cilium/cilium/pull/36098
https://github.com/cilium/cilium/pull/36099

The 1.14 one should be merged first, then we can re-run 1.15 a few times to get some data.

**Comment by gandro** [comment]
Since the above PRs were merged (thanks Timo!), there have been no tail-call failures on `ci-e2e-upgrade` in v1.15. I'm closing this as completed.

**If you are reading this because you observed "Missed tail call" in CI, please consider creating a new issue**, since it likely has a different root cause than the two that were solved here.

In particular, make sure to write down:

1. **What branch did the failure occur?**
2. **What particular test did fail?** (i.e. test number in ci-e2e-upgrade)
1. **What image version did the failure occur in?** (this is in particular important since upgrade tests run two versions of Cilium, so a broken v1.14 image might break the v1.15 workflow)
3. **When did the failures start to occur?**

All those things are very important to keep track of, since it _a missed tail call can have may root causes_ and it's important to not mix up the different causes.

Especially if you see it occurring in a PR of yours and the above indicators look different than the ones reported so far (e.g. new branch or new failing test), then it is possible that your PR is introducing a regression (this was the case for some failure reported earlier in this issue).
